core/src/main/java/com/linecorp/armeria/common/PublisherBasedHttpResponse.java

```diff
@@ -29,6 +30,13 @@ static PublisherBasedHttpResponse from(ResponseHeaders headers, Publisher<? exte
         return new PublisherBasedHttpResponse(new PrependingPublisher<>(headers, publisher));
     }
 
+    static PublisherBasedHttpResponse from(ResponseHeaders headers,
+                                           Publisher<? extends HttpObject> publisher,
+                                           HttpHeaders trailers) {
+        return new PublisherBasedHttpResponse(new AppendingPublisher<>(
+                new PrependingPublisher<>(headers, publisher), trailers));
```

Comment:
This makes me think what we actually want is to evolve `PrependingPublisher` into `SurroundingPublisher` that allows us to add 'head' and 'tail' element optionally:

```
new SurroundingPublisher<>(headers, publisher, trailers);
new SurroundingPublisher<>(headers, publisher, null); // Prepending only
new SurroundingPublisher<>(null, publisher, trailers); // Appending only
```

We could convert the current boolean-based state machine (`firstSent`) into an enum based one:
```java
enum State {
  SENDING_HEAD, // This becomes the initial state if head is non-null.
  SENDING_BODY, // This becomes the initial state if head is null.
  SENDING_TAIL, // This state is skipped if tail is null.
  COMPLETE
}
```

What do you think?

Reply:
- Option 1: keep `Appending&PrependingPublisher` separately
- Option 2: Merge to `SurroundingPublisher` that can handle both append and prepend

@minwoox , @ikhoon , @jrhee17 , WDYT? I think `option 2` looks better 🤔

If we choose `option 2`, I'll update it on this PR~!

Reply:
SurroundingPublisher looks efficient because we can remove a StreamMessage chain. 

By the way, if we use a function to create the last element, it can be useful when users want to dynamically add an element depending on the last signal of the upstream.

We can think about the following cases.
- If the upstream emits an error via `onError()`:
  - Convert the error into the last message and call `onComplete()`
  - Ignore the error and just call `onComplete()`
  - Bypass the error to the downstream
- If the stream emits an `onComplete()` signal
  - Append the last element and call `onComplete()`
  - Bypass the signal to the downstream

This is A prototype for the API. It needs a better name though.
```java
StreamMessage stream = StreamMessage.of(...);
stream.endWith(cause -> {
    if (cause == null) { // onComplete() is called by the upstream
        return obj;
    } else {
        return recover(cause);
    }
});
```

The dynamically appendable `Publisher` could be used to implement the WebSocket service. 
https://github.com/line/armeria/blob/19151268fe15721edaff0ef011d90e6e811dcffd/core/src/main/java/com/linecorp/armeria/server/websocket/WebSocketResponseSubscriber.java#L87-L103

Reply:
I'll update this PR to `SurroundingPublisher` soon~! thanks for nice suggestion 🙇

Reply:
```
StreamMessage stream = StreamMessage.of(...);
stream.endWith(cause -> {
```
@ikhoon nim, should I add `endWith` on `SurroundingPublisher` or `StreamMessage`?

Reply:
`endWith` method should be added to `StreamMessage`.
The implementation of `endWith` will use `SurroundingPublisher`.
For example:
```java
interface StreamMessage<T> {

   <T> StreamMessage<T> endWith(Function<@Nullable Throwable, ? extends T> finalizer) {
       // Conditionally appends an element using finalizer
       return new SurroundingPublisher(null, this, finalizer);
   }
}
```

Reply:
> endWith method should be added to StreamMessage.

aha~ then may I handle it with next PR?
If you make related issue, then I'll assign it to me 🙇

Reply:
It can cause some changes to the implementation so, I thought, it would be better to consider it when you initially design the code. But if you want, it can handle it separately. I will create a issue for it.

Reply:
#4816 has been created. 😉

Reply:
Let us know if you are ready to review.

Reply:
> It can cause some changes to the implementation so, I thought, it would be better to consider it when you initially design the code

aha.. I understand you intention! but I think with `endWith`, this PR may too big to review 😅
please review this PR~! I'll add `endWith` right after this PR is merged. (to include in the same version!)